### PR TITLE
Extend SCC transformation to loops in driver

### PR DESCRIPTION
**Pull request by MichaelSt98**
to apply loop flip, vector and seq annotations at driver level too

* introducing two new Loki pragmas
    * `!$loki separator` - for extract vector sections in order to have assignments in the correct position
    * `!$loki driver-loop` - for recognising a driver loop without a call in it

### CODE REVIEW

File: transformations/transformations/single_column_coalesced_vector.py [hunk 1]
```diff
@@ -193,6 +199,67 @@ def process_kernel(self, routine):
...
+            new_driver_loop = driver_loop.clone(body=new_driver_loop)
+            new_driver_loops.append(new_driver_loop)
+            driver_loop_map[driver_loop] = new_driver_loop
+
+        routine.body = Transformer(driver_loop_map).visit(routine.body)
+
+        driver_loop_map = {}
+        for driver_loop in new_driver_loops:
+            # Extract vector-level compute sections from the kernel
+            sections = self.extract_vector_sections(driver_loop.body, self.horizontal)
```
Comment (reuterbal) [inline transformations/transformations/single_column_coalesced_vector.py:256]: Could we apply this directly to the `driver_loop` before putting it into the map that is then applied via the transformer? This seems like an unnecessary two-step replacement procedure for the same loops.

**Review by reuterbal — CHANGES_REQUESTED** [review]
This is a great find and useful addition, many thanks!

While I agree with the changes in principle, I think the implementation could be improved in a number of ways to streamline the control flow and reduce code duplication, unless I misunderstood some of the changes. See inline comments for more details.

File: transformations/transformations/single_column_coalesced_vector.py [hunk 2]
```diff
@@ -73,8 +73,11 @@ def extract_vector_sections(cls, section, horizontal):
         _scope_node_types = (ir.Loop, ir.Conditional, ir.MultiConditional)
 
         # Identify outer "scopes" (loops/conditionals) constrained by recursive routine calls
-        separator_nodes = []
         calls = FindNodes(ir.CallStatement).visit(section)
+        pragmas = [pragma for pragma in FindNodes(ir.Pragma).visit(section) if pragma.keyword.lower() == "loki" and
```
Comment (awnawab) [inline transformations/transformations/single_column_coalesced_vector.py:77]: If I have understood the use-case for `!$loki separator` pragmas correctly, I think the same functionality can be achieved by triggering the `trim_vector_sections` option on the driver loops too.

The only difficulty there is calling the dataflow analysis on `driver_loop` since loops are frozen ir Nodes and the dataflow analysis attacher wants to assign to `driver_loop.body`.

I did a quick hack and simply ran the dataflow analysis attacher as an in-place transform that doesn't return anything. That seems to work; the relevant `_live/_uses/_defines_symbols` sets were updates correctly.

Long story short: if you rely on this leave this in, but I'll probably look into replacing the `!$loki separator` functionality with `trim_vector_sections` once I've had the chance to understand the `DataflowAnalysisAttacher` a bit better.

**Review by reuterbal — APPROVED** [review]
Many thanks, I think it looks much cleaner now. Very valuable addition, GTG from my side!

File: transformations/transformations/single_column_coalesced_vector.py [hunk 3]
```diff
@@ -193,6 +199,41 @@ def process_kernel(self, routine):
...
+        for loop in driver_loops:
+            loop_map = {}
+            for l in FindNodes(ir.Loop).visit(loop.body):
+                if l.variable == self.horizontal.index:
+                    loop_map[l] = l.body
+            new_driver_loop = Transformer(loop_map).visit(loop.body)
+            new_driver_loop = loop.clone(body=new_driver_loop)
+            sections = self.extract_vector_sections(new_driver_loop.body, self.horizontal)
+            if self.trim_vector_sections:
+                sections = self.get_trimmed_sections(routine, self.horizontal, sections)
```
Comment (awnawab) [inline transformations/transformations/single_column_coalesced_vector.py:231]: Have you tested this with `trimmed_vector_sections` enabled? I believe in the current guise of the dataflow analysis this will give a runtime error. #158 tries to update the dataflow analysis to fix this.

Comment (awnawab) [reply transformations/transformations/single_column_coalesced_vector.py:231]: Once that is in, I believe you should be able to remove `!$loki separator` pragmas.

Comment (MichaelSt98) [reply transformations/transformations/single_column_coalesced_vector.py:231]: Tested it with and without `self.get_trimmed_sections(...)`. However, (to be honest) haven't checked whether it actually makes a difference (for the code parts I used it for). Regarding the removal `!$loki separator` we should probably make this another PR once confirmed that it works (considering possible edge cases)?!

**Review by mlange05 — COMMENTED** [review]
All in all this looks good and brings in the required functionality. My main question/comment would be if we could use the `block_dim` dimension be used instead of introducing a dedicated `!$loki driver-loop` pragma? I'm happy to go ahead with this if there's a corner-case I've missed, so just wondering what that looks like.

I've also highlighted a few cases where using the "one single" definition of the driver loop should be used, instead of case-specific selectors (happy to discuss if this is too complicated in places). Otherwise I'm happy with this. 

For the `!$loki separator` pragma, I can definitely see how this is useful for development beyond the use equivalent to the "trim_vector_section" feature. This should always be a dev-backdoor though, with the section-trimmer the preferred option.

So, for now I've left this as a comment to see if we could improve some of the small details of this. I notice that this also requires a rebase now; and I'm running the ec-phys regression now.

File: transformations/transformations/single_column_coalesced.py [hunk 4]
```diff
@@ -190,6 +190,59 @@ def resolve_vector_dimension(cls, routine, loop_variable, bounds):
...
+        for loop in FindNodes(ir.Loop).visit(routine.body):
+            if loop in nested_driver_loops:
+                continue
+
+            if not cls.is_driver_loop(loop, targets):
+                continue
+
+            driver_loops.append(loop)
+            loops = FindNodes(ir.Loop).visit(loop.body)
+            nested_driver_loops.extend(loops)
```
[the remark this answers is not in the document]
Comment (MichaelSt98) [reply transformations/transformations/single_column_coalesced.py:243]: If I'm not mistaken your proposed alternative would not work...

`cls.is_driver_loop(...)` checks either for the driver loop pragma or for calls that are within `target`. `FindNodes(ir.Loop).visit(routine.body)` returns all the loops (whether nested or not) in the order of their occurrence. That's what we rely on in order to get only the outermost loop and not the corresponding nested loops ...

E.g., 

```fortran
DO J=1,10
 DO K=1,10
  CALL KERNEL(...)
 END DO
END DO
```

The current implementation returns only the `J` loop, whereas your proposed alternative would return both loops.

File: transformations/transformations/single_column_coalesced.py [hunk 5]
```diff
@@ -354,15 +407,20 @@ def kernel_annotate_sequential_loops_openacc(cls, routine, horizontal):
...
+        block_dim_index = None if block_dim is None else block_dim.index
         with pragmas_attached(routine, ir.Loop):
 
             for loop in FindNodes(ir.Loop).visit(routine.body):
                 # Skip loops explicitly marked with `!$loki/claw nodep`
                 if loop.pragma and any('nodep' in p.content.lower() for p in as_tuple(loop.pragma)):
                     continue
 
-                if loop.variable != horizontal.index:
+                if loop.variable != horizontal.index and loop.variable != block_dim_index and loop not in ignore:
```
[the remark this answers is not in the document]
Comment (MichaelSt98) [reply transformations/transformations/single_column_coalesced.py:423]: I agree. To be honest I thought we would determine "driver loops" via `block_dim`. However, currently we determine "driver loops" via calls to subroutines that are within `target` (and now additionally via the driver loop pragma). I don't know whether there is (or was) a reason for doing it via calls that are within `target` instead of using `block_dim`. I am open for discussion. However, switching to determining "driver loops" via `block_dim` would need some extensive testing to ensure we don't break anything ...

File: transformations/transformations/single_column_coalesced.py [hunk 6]
```diff
@@ -577,9 +632,12 @@ def annotate_driver(cls, directive, driver_loop, kernel_loop, block_dim, column_
 
             # Annotate vector loops with OpenACC pragmas
             if kernel_loop:
-                kernel_loop._update(pragma=ir.Pragma(keyword='acc', content='loop vector'))
+                for loop in kernel_loop:
+                    loop._update(pragma=(ir.Pragma(keyword='acc', content='loop vector'),))
 
-            if driver_loop.pragma is None:
+            if driver_loop.pragma is None or (len(driver_loop.pragma) == 1 and
```
[the remark this answers is not in the document]
Comment (MichaelSt98) [reply transformations/transformations/single_column_coalesced.py:638]: Yes, but the additional check is for a different reason. 
The problem is that I've added the driver loop pragma and with the former/original logic this means that the driver loop has a pragma (thus `driver_loop.pragma` is None evaluates to `False`) but this pragma is not a `acc data` statement and consequently the `acc parallel loop gang ...` would not be inserted ... 
I think we should improve the logic here. However, I think for now it's reasonable to do it like this and improve it whenever we add OpenMP pragmas (as alternative to OpenACC pragmas) ?!

**Comment by MichaelSt98** [comment]
@mlange05 Please let me know what you think about my comments and afterwards I'll address the "minor" things as well. Thanks for having a look and thank you for the comments!

File: transformations/transformations/single_column_coalesced.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (mlange05) [reply transformations/transformations/single_column_coalesced.py:243]: Ahh ok, cool, makes sense.

File: transformations/transformations/single_column_coalesced.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (mlange05) [reply transformations/transformations/single_column_coalesced.py:423]: Ok, agreed, and good point! Let's leave it like this for now, and I'll file a reminder issue that we want to clean-up in a separate PR with the extensive testing.

File: transformations/transformations/single_column_coalesced.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (mlange05) [reply transformations/transformations/single_column_coalesced.py:638]: Yes, again, good point, and fully agreed. This will need a little logic reshuffling, but it's reasonable to attepmt this later (and possibly once OpenMP-offload is needed for this.

**Review by mlange05 — APPROVED** [review]
@MichaelSt98 Happy with your comments and explanations. I'll file a separate issue for tracking the "identify via `block_dim`" appraoch, but as you rightly pointed out this needs some though and separate testing. 

GTG from me once rebased and minor points addressed. :shipit:  Well done and many thanks!!!

File: transformations/transformations/single_column_coalesced.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (reuterbal) [reply transformations/transformations/single_column_coalesced.py:638]: This sounds like another use-case where better API to work with pragmas might be useful. We should improve our utilities for that.

**Comment by mlange05** [comment]
Argh! Seems this change breaks EC-physics regression - my current setup at least. Will need to hold this until I can figure out what's wrong here. Apologies @MichaelSt98 , I'll try and get this done ASAP.

**Comment by mlange05** [comment]
Ok, found the issue and can now confirm that EC-physics regression works. If vector-trimming is not enabled, `!$loki separator` pragmas are needed in the block-loops of all driver routines, to avoid the vector loops being hoisted to far and interfering with the bounds calculations (`KIDIA`/`KFDIA`). This fixes the local problem for now, and I'll investigate vector-section trimming separately.

This, however, now means we are GTG for merge! :tada: :shipit:
